src/librustdoc/test.rs

```diff
@@ -308,49 +228,62 @@ fn run_test(
-        stderr: Some(data.clone()),
-        crate_name: None,
-        lint_caps: Default::default(),
-    };
+    let current_dir = env::current_dir().expect("rustdoc needs to create a tempfile");
+    let input_dir = options.input.parent().unwrap_or(&current_dir);
+    let input_file = TempFileBuilder::new()
+                      .prefix("tempRustDocTestInput")
+                      .suffix(".rs")
+                      .tempfile_in(input_dir)
+                      .expect("rustdoc needs to create a tempfile");
+    std::fs::write(input_file.path(), test).expect("could write out test sources");
+
+    let mut compiler = Command::new(std::env::current_exe().unwrap().with_file_name("rustc"));
+    for cfg in &options.cfgs {
+        compiler.arg("--cfg").arg(&cfg);
+    }
+    if let Some(sysroot) = options.maybe_sysroot {
+        compiler.arg("--sysroot").arg(sysroot);
+    }
+    compiler.arg("--edition").arg(&edition.to_string());
+    compiler.arg(input_file.path()).current_dir(outdir.path());
+    compiler.arg("-o").arg(&output_file);
+    if as_test_harness {
+        compiler.arg("--test");
+    }
+    for lib_str in &options.lib_strs {
+        compiler.arg("-L").arg(&lib_str);
+    }
+    for extern_str in &options.extern_strs {
+        compiler.arg("--extern").arg(&extern_str);
+    }
+    for codegen_options_str in &options.codegen_options_strs {
+        compiler.arg("-C").arg(&codegen_options_str);
+    }
+    if let Some(linker) = options.linker {
+        compiler.arg(&format!("-C linker={:?}", linker));
+    }
 
-    let compile_result = panic::catch_unwind(AssertUnwindSafe(|| {
-        interface::run_compiler(config, |compiler| {
-            if no_run {
-                compiler.global_ctxt().and_then(|global_ctxt| global_ctxt.take().enter(|tcx| {
-                    tcx.analysis(LOCAL_CRATE)
-                })).ok();
-            } else {
-                compiler.compile().ok();
-            };
-            compiler.session().compile_status()
-        })
-    })).map_err(|_| ()).and_then(|s| s.map_err(|_| ()));
+    let cmd = format!("{:?}", compiler);
+    let output = compiler.output().unwrap();
 
-    match (compile_result, compile_fail) {
-        (Ok(()), true) => {
+    match (output.status.success(), compile_fail) {
+        (true, true) => {
             return Err(TestFailure::UnexpectedCompilePass);
         }
-        (Ok(()), false) => {}
-        (Err(_), true) => {
+        (true, false) => {}
+        (false, true) => {
             if !error_codes.is_empty() {
-                let out = String::from_utf8(data.lock().unwrap().to_vec()).unwrap();
+                let out = String::from_utf8(output.stderr).unwrap();
                 error_codes.retain(|err| !out.contains(err));
 
                 if !error_codes.is_empty() {
                     return Err(TestFailure::MissingErrorCodes(error_codes));
                 }
             }
         }
-        (Err(_), false) => {
+        (false, false) => {
```

Comment:
how shall the fail printout look thinks that it is a bit messy now

Reply:
I think this is fine for now, a future cleanup PR might be in order but that doesn't need to happen in this PR.

Reply:
ok